src/internal/obj/local_client.go

```diff
@@ -1,126 +1,163 @@
 package obj
 
 import (
+	"bytes"
 	"context"
+	"encoding/base64"
+	"fmt"
 	"io"
 	"os"
 	"path/filepath"
 	"strings"
 
 	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
 	"github.com/pachyderm/pachyderm/v2/src/internal/pacherr"
-	"github.com/pachyderm/pachyderm/v2/src/internal/tracing"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/sirupsen/logrus"
 )
 
 // NewLocalClient returns a Client that stores data on the local file system
-func NewLocalClient(root string) (Client, error) {
-	if err := os.MkdirAll(root, 0755); err != nil {
-		return nil, errors.EnsureStack(err)
+func NewLocalClient(rootDir string) (Client, error) {
+	c, err := newFSClient(rootDir)
+	if err != nil {
+		return nil, err
 	}
-	var c Client
-	c = &localClient{filepath.Clean(root)}
 	if monkeyTest {
 		c = &monkeyClient{c}
 	}
 	return newUniformClient(c), nil
 }
 
-type localClient struct {
-	root string
+type fsClient struct {
+	rootDir string
 }
 
-func (c *localClient) normPath(path string) string {
-	path = filepath.Clean(path)
-	return filepath.Join(c.root, path)
+func newFSClient(rootDir string) (Client, error) {
+	c := &fsClient{
+		rootDir: filepath.Clean(rootDir),
+	}
+	if c.rootDir == "" || c.rootDir == "/" || c.rootDir == "." {
+		panic("you probably didn't want to set the local client's root path to " + c.rootDir)
+	}
+	if err := c.init(); err != nil {
+		return nil, err
+	}
+	return c, nil
 }
 
-func (c *localClient) Put(_ context.Context, path string, r io.Reader) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	fullPath := c.normPath(path)
-
-	// Create the directory since it may not exist
-	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
+func (c *fsClient) Put(ctx context.Context, name string, r io.Reader) (retErr error) {
+	staging := c.stagingPathFor(name)
+	final := c.finalPathFor(name)
+	f, err := os.Create(staging)
+	if err != nil {
 		return err
 	}
-	file, err := os.Create(fullPath)
-	if err != nil {
+	defer c.closeFile(&retErr, f)
```

Comment:
I think we want to close and remove it upon error.

Reply:
This will error if the Put can't be completed.  If the staging file can't be cleaned up, it will log the error, but the Put still succeeded so we shouldn't return an error.  We clean up the staging area during initialization, so we won't be leaking staging files indefinitely.

Reply:
I think we should make an attempt at removing the file from the staging area if an error occurs, rather than relying on a restart and having a leak in between.

Reply:
Added `removeFile` similar to `closeFile` to handle this.